proposals/eligibility_validator.go

```diff
@@ -74,35 +68,44 @@ func (v *Validator) CheckEligibility(ctx context.Context, ballot *types.Ballot,
 	if len(ballot.EligibilityProofs) == 0 {
 		return false, fmt.Errorf("empty eligibility list is invalid (ballot %s)", ballot.ID())
 	}
-	owned, err := v.cdb.GetAtxHeader(ballot.AtxID)
-	if err != nil {
-		return false, fmt.Errorf("failed to load atx %v: %w", ballot.AtxID, err)
-	}
-	if owned.TargetEpoch() != ballot.Layer.GetEpoch() {
-		return false, fmt.Errorf("%w: ATX target epoch (%v), ballot publication epoch (%v)",
-			errTargetEpochMismatch, owned.TargetEpoch(), ballot.Layer.GetEpoch())
-	}
-	if ballot.SmesherID != owned.NodeID {
-		return false, fmt.Errorf("%w: public key (%v), ATX node key (%v)", errPublicKeyMismatch, ballot.SmesherID.String(), owned.NodeID)
-	}
-	nonce, err := v.nonceFetcher.VRFNonce(ballot.SmesherID, ballot.Layer.GetEpoch())
-	if err != nil {
-		return false, fmt.Errorf("no vrf nonce for %v in epoch %v: %w", ballot.SmesherID, ballot.Layer.GetEpoch(), err)
-	}
-	var data *types.EpochData
-	if ballot.EpochData != nil && ballot.Layer.GetEpoch() == v.clock.CurrentLayer().GetEpoch() {
-		var err error
-		data, err = v.validateReference(ballot, actives, owned)
+	var atx *cache.ATXData
+	if epoch := ballot.Layer.GetEpoch(); v.cache.IsEvicted(epoch) {
+		owned, err := atxs.Get(v.db, ballot.AtxID)
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to load atx %v: %w", ballot.AtxID, err)
 		}
-	} else {
-		var err error
-		data, err = v.validateSecondary(ballot, owned)
+		if owned.TargetEpoch() != epoch {
+			return false, fmt.Errorf("atx and ballot epochs mismatch. atx %d/%s is not from %d/%s", owned.TargetEpoch(), ballot.AtxID.ShortString(), epoch, ballot.SmesherID.ShortString())
+		}
+		if ballot.SmesherID != owned.SmesherID {
+			return false, fmt.Errorf("atx and ballot key mismatch: public key (%v), ATX node key (%v)", ballot.SmesherID.String(), owned.NodeID)
+		}
+		nonce, err := atxs.VRFNonce(v.db, ballot.SmesherID, epoch)
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("no vrf nonce for %v in epoch %v: %w", ballot.SmesherID, ballot.Layer.GetEpoch(), err)
+		}
+		atx = cache.ToATXData(owned.ToHeader(), nonce, false)
```

Comment:
why is the malicious bit always false here? there should be a lookup in db for that.

Reply:
i will change this data to make it clear that bit is unused